ast/src/parsed/mod.rs

```diff
@@ -650,25 +613,42 @@ pub trait SourceReference {
 }
 
 macro_rules! impl_source_reference {
+    // Version for types with generic parameter
+    ($enum:ident<$generic:ident>, $($variant:ident),*) => {
+        impl<$generic> SourceReference for $enum<$generic> {
+            fn source_reference(&self) -> &SourceRef {
```

Comment:
Maybe extract the lines starting here into a macro `impl_source_reference_inner`?